### ISSUE TITLE
mosquitto_loop_read exits and reconnects if a topic that does not UTF8 validate is received, causing a never ending reconnect loop

### ISSUE THREAD

**Issue by NiclasLindgren** [eclipse/mosquitto#2168]
In int handle__publish(struct mosquitto *mosq) There is packet__read_string that returns (correctly) an error on malformed UTF8, but that should not make the connection reset I believe?

**Comment by ralight** [comment]
Which broker are you using that is sending you malformed UTF-8 packets? :)

**Comment by NiclasLindgren** [comment]
mosquitto =) (version 1.4.14)

**Comment by NiclasLindgren** [comment]
I noticed while upgrading to 2.0.10 and trying out the mosquitto_sub client, suddently I didn't get a flow of messages, just a constant loop of retained messages (up to the "broken" topic) over and over again.

I suspect using the 2.0.10 broker would reject the topic all together, it seems like it in the code, but it almost seems like the broker would also disconnect, don't fully grasp the code paths without debugging

**Comment by NiclasLindgren** [comment]
![image](https://user-images.githubusercontent.com/6535699/114078232-f955a780-98a8-11eb-906b-cd5cfbb6f949.png)

**Comment by NiclasLindgren** [comment]
If I understand the code properly, in broker mode, the broker will disconnect that client on an invalid topic in mux_poll.c->loop_handle_reads_writes

			do{
				rc = packet__read(context);
				if(rc){
					do_disconnect(context, rc);
					continue;
				}
			}while(SSL_DATA_PENDING(context));

So that is likely better, still a bit "weird" that the mosquitto client (or any client using the lib) will get a silent disconnect/reconnect

**Comment by ralight** [comment]
Thanks. Version 1.4 wasn't spec compliant so it didn't verify topics were UTF-8. The broker does disconnect clients that send malformed topics (since 1.5) and is required to do so by the spec. The client library produces an error if you try to publish with malformed UTF-8, and the on_disconnect() callback will be called if you receive a bad message from the broker. It's not ideal, but there are mechanisms in place to help a little.

**Comment by ralight** [comment]
I'd always argue in favour of getting that old broker updated, but you're right we have to be a bit pragmatic as well.

**Comment by NiclasLindgren** [comment]
I agree about upgrading, I am trying to do that but hit a few snags, but in the end often we don't control the broker we are just the client. In this case I think it is EMQ
